week9/answers.sql

```diff
@@ -1,3 +1,117 @@
 -- Create your tables, views, functions and procedures here!
 CREATE SCHEMA social;
 USE social;
+
+
+CREATE TABLE users (
+    user_id INT UNSIGNED AUTO_INCREMENT PRIMARY KEY,
+    first_name VARCHAR(50),
+    last_name VARCHAR(50),
+    email VARCHAR(100),
+    created_on TIMESTAMP DEFAULT CURRENT_TIMESTAMP
+);
+
+CREATE TABLE sessions (
+    session_id INT UNSIGNED AUTO_INCREMENT PRIMARY KEY,
+    user_id INT UNSIGNED,
+    created_on TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    updated_on TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+    FOREIGN KEY (user_id) REFERENCES users(user_id)
+);
+
+CREATE TABLE friends (
+    user_friend_id INT UNSIGNED AUTO_INCREMENT PRIMARY KEY,
+    user_id INT UNSIGNED,
+    friend_id INT UNSIGNED,
+    FOREIGN KEY (user_id) REFERENCES users(user_id),
+    FOREIGN KEY (friend_id) REFERENCES users(user_id)
+);
+
+CREATE TABLE posts (
+    post_id INT UNSIGNED AUTO_INCREMENT PRIMARY KEY,
+    user_id INT UNSIGNED,
+    created_on TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    updated_on TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+    content TEXT,
+    FOREIGN KEY (user_id) REFERENCES users(user_id)
+);
+
+CREATE TABLE notifications (
+    notification_id INT UNSIGNED AUTO_INCREMENT PRIMARY KEY,
+    user_id INT UNSIGNED,
+    post_id INT UNSIGNED,
+    FOREIGN KEY (user_id) REFERENCES users(user_id),
+    FOREIGN KEY (post_id) REFERENCES posts(post_id)
+);
+
+CREATE VIEW notification_posts AS
+SELECT n.user_id AS notification_user_id, 
+       u.first_name, 
+       u.last_name, 
+       p.post_id, 
+       p.content
+FROM notifications n
+LEFT JOIN users u ON n.user_id = u.user_id
+LEFT JOIN posts p ON n.post_id = p.post_id;
+
+DELIMITER //
+CREATE TRIGGER new_user_notification AFTER INSERT ON users
+FOR EACH ROW
+BEGIN
+    DECLARE message VARCHAR(255);
+    SET message = CONCAT(NEW.first_name, ' ', NEW.last_name, ' just joined!');
+
+    INSERT INTO notifications (user_id, post_id)
```

Comment:
Looking for a CURSOR in this solution. If you have 1 billion users, your server would likely run out of memory while performing this query.

Otherwise, the logic for this trigger isn't accurate. You'll need to create a post and add a notification for each user.